File: app/modules/processes/schemas.py
from __future__ import annotations

from datetime import datetime

from app.modules.shared.schemas import DashboardModel


class ProcessTreeNode(DashboardModel):
    pid: int
    ppid: int
    pgid: int
    sid: int
    state: str
    status_label: str
    started_at: datetime | None = None
    elapsed_seconds: int | None = None
    command_name: str
    command_kind: str
    display_command: str
    role: str
    task_label: str | None = None
    lifecycle_status: str = "running"
    ended_at: datetime | None = None
    read_at: datetime | None = None
    cwd: str | None = None
    repo_path: str | None = None
    children: list[ProcessTreeNode]


class ProcessTreesResponse(DashboardModel):
    collected_at: datetime
    poll_interval_seconds: int
    total_trees: int
    total_processes: int
    trees: list[ProcessTreeNode]


class ProcessTreesClearUnreadResponse(DashboardModel):
    cleared_count: int

File: frontend/src/features/processes/schemas.ts
import { z } from "zod";

export type ProcessTreeNode = {
  pid: number;
  ppid: number;
  pgid: number;
  sid: number;
  state: string;
  statusLabel: string;
  startedAt: string | null;
  elapsedSeconds: number | null;
  commandName: string;
  commandKind: string;
  displayCommand: string;
  role: string;
  taskLabel: string | null;
  lifecycleStatus: "running" | "ended";
  endedAt: string | null;
  readAt: string | null;
  cwd: string | null;
  repoPath: string | null;
  children: ProcessTreeNode[];
};

export const ProcessTreeNodeSchema: z.ZodType<ProcessTreeNode> = z.lazy(() =>
  z.object({
    pid: z.number(),
    ppid: z.number(),
    pgid: z.number(),
    sid: z.number(),
    state: z.string(),
    statusLabel: z.string(),
    startedAt: z.string().nullable(),
    elapsedSeconds: z.number().nullable(),
    commandName: z.string(),
    commandKind: z.string(),
    displayCommand: z.string(),
    role: z.string(),
    taskLabel: z.string().nullable(),
    lifecycleStatus: z.enum(["running", "ended"]).default("running"),
    endedAt: z.string().nullable().default(null),
    readAt: z.string().nullable().default(null),
    cwd: z.string().nullable(),
    repoPath: z.string().nullable(),
    children: z.array(ProcessTreeNodeSchema),
  }),
);

export const ProcessTreesResponseSchema = z.object({
  collectedAt: z.string(),
  pollIntervalSeconds: z.number(),
  totalTrees: z.number(),
  totalProcesses: z.number(),
  trees: z.array(ProcessTreeNodeSchema),
});

export type ProcessTreesResponse = z.infer<typeof ProcessTreesResponseSchema>;

export const ProcessTreesClearUnreadResponseSchema = z.object({
  clearedCount: z.number().int().nonnegative(),
});

export type ProcessTreesClearUnreadResponse = z.infer<typeof ProcessTreesClearUnreadResponseSchema>;
